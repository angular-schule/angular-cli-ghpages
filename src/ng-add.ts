import { workspaces } from '@angular-devkit/core';
import {
  SchematicContext,
  SchematicsException,
  Tree
} from '@angular-devkit/schematics';
import { createHost } from './utils';

interface NgAddOptions {
  project: string;
}

export const ngAdd = (options: NgAddOptions) => async (
  tree: Tree,
  _context: SchematicContext
) => {
  const host = createHost(tree);
  const { workspace } = await workspaces.readWorkspace('/', host);

  if (!options.project) {
    if (workspace.projects.size === 1) {
      // If there is only one project, return that one.
      options.project = Array.from(workspace.projects.keys())[0];
    } else {
      throw new SchematicsException(
        'There is more than one project in your workspace. Please select it manually by using the --project argument.'
      );
    }
  }

  const project = workspace.projects.get(options.project);
  if (!project) {
    throw new SchematicsException(
      'The specified Angular project is not defined in this workspace'
    );
  }

  if (project.extensions.projectType !== 'application') {
    throw new SchematicsException(
      `Deploy requires an Angular project type of "application" in angular.json`
    );
  }

  if (!project.targets.get('build')?.options?.outputPath) {
    throw new SchematicsException(
      `Cannot read the output path (architect.build.options.outputPath) of the Angular project "${options.project}" in angular.json`
    );
  }

  project.targets.add({
    name: 'deploy',
    builder: 'angular-cli-ghpages:deploy',
    options: {}
  });

  workspaces.writeWorkspace(workspace, host);
  return tree;
};
