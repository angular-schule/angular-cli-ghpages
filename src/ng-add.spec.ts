import { SchematicContext, Tree } from '@angular-devkit/schematics';

import { ngAdd } from './ng-add';

const PROJECT_NAME = 'THEPROJECT';
const PROJECT_ROOT = 'PROJECTROOT';
const OTHER_PROJECT_NAME = 'OTHERPROJECT';

describe('ng-add', () => {
  describe('generating files', () => {
    let tree: Tree;

    beforeEach(() => {
      tree = Tree.empty();
      tree.create('angular.json', JSON.stringify(generateAngularJson()));
    });

    it('generates new files if starting from scratch', async () => {
      const result = await ngAdd({
        project: PROJECT_NAME
      })(tree, {} as SchematicContext);

      const actual = result.read('angular.json')!.toString();
      expect(prettifyJSON(actual)).toMatchSnapshot();
    });

    it('overrides existing files', async () => {
      const tempTree = await ngAdd({
        project: PROJECT_NAME
      })(tree, {} as SchematicContext);

      const result = await ngAdd({
        project: OTHER_PROJECT_NAME
      })(tempTree, {} as SchematicContext);

      const actual = result.read('angular.json')!.toString();

      expect(prettifyJSON(actual)).toMatchSnapshot();
    });
  });

  describe('project selection', () => {

    it('should select the first project if there is only one', async () => {
      const tree = Tree.empty();
      const angularJSON = generateAngularJson();
      delete (angularJSON as any).projects[PROJECT_NAME]; // delete one project so that one is left
      tree.create('angular.json', JSON.stringify(angularJSON));

      const resultTree = await ngAdd({ project: '' })(
        tree,
        {} as SchematicContext
      );

      const resultConfig = readJSONFromTree(resultTree, 'angular.json');
      expect(
        resultConfig.projects[OTHER_PROJECT_NAME].architect.deploy
      ).toBeTruthy();
    });
  });

  describe('error handling', () => {
    it('should fail if there are multiple projects in workspace and project is not explicitly defined', async () => {
      const tree = Tree.empty();
      const angularJSON = generateAngularJson();
      tree.create('angular.json', JSON.stringify(angularJSON));

      await expect(
        ngAdd({ project: '' })(tree, {} as SchematicContext)
      ).rejects.toThrowError(
        'There is more than one project in your workspace. Please select it manually by using the --project argument.'
      );
    });

    it('should throw if angular.json not found', async () => {
      await expect(
        ngAdd({ project: PROJECT_NAME })(Tree.empty(), {} as SchematicContext)
      ).rejects.toThrowError('Unable to determine format for workspace path.');
    });

    it('should throw if angular.json can not be parsed', async () => {
      const tree = Tree.empty();
      tree.create('angular.json', 'hi');

      await expect(
        ngAdd({ project: PROJECT_NAME })(tree, {} as SchematicContext)
      ).rejects.toThrowError('Invalid workspace file - expected JSON object.');
    });

    it('should throw if specified project does not exist', async () => {
      const tree = Tree.empty();
      tree.create('angular.json', JSON.stringify({ version: 1, projects: {} }));

      await expect(
        ngAdd({ project: PROJECT_NAME })(tree, {} as SchematicContext)
      ).rejects.toThrowError(
        'The specified Angular project is not defined in this workspace'
      );
    });

    it('should throw if specified project is not application', async () => {
      const tree = Tree.empty();
      tree.create(
        'angular.json',
        JSON.stringify({
          version: 1,
          projects: {
            [PROJECT_NAME]: { projectType: 'invalid', root: PROJECT_NAME }
          }
        })
      );

      await expect(
        ngAdd({ project: PROJECT_NAME })(tree, {} as SchematicContext)
      ).rejects.toThrowError(
        'Deploy requires an Angular project type of "application" in angular.json'
      );
    });

    it('should throw if app does not have architect configured', async () => {
      const tree = Tree.empty();
      tree.create(
        'angular.json',
        JSON.stringify({
          version: 1,
          projects: {
            [PROJECT_NAME]: { projectType: 'application', root: PROJECT_NAME }
          }
        })
      );

      await expect(
        ngAdd({
          project: PROJECT_NAME
        })(tree, {} as SchematicContext)
      ).rejects.toThrowError(
        'Cannot read the output path (architect.build.options.outputPath) of the Angular project "THEPROJECT" in angular.json'
      );
    });
  });
});

function prettifyJSON(json: string) {
  return JSON.stringify(JSON.parse(json), null, 2);
}

function readJSONFromTree(tree: Tree, file: string) {
  return JSON.parse(tree.read(file)!.toString());
}

function generateAngularJson() {
  return {
    version: 1,
    projects: {
      [PROJECT_NAME]: {
        projectType: 'application',
        root: PROJECT_ROOT,
        architect: {
          build: {
            options: {
              outputPath: 'dist/' + PROJECT_NAME
            }
          }
        }
      },
      [OTHER_PROJECT_NAME]: {
        projectType: 'application',
        root: PROJECT_ROOT,
        architect: {
          build: {
            options: {
              outputPath: 'dist/' + OTHER_PROJECT_NAME
            }
          }
        }
      }
    }
  };
}
