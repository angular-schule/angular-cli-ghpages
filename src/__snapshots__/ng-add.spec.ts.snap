// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ng-add generating files generates new files if starting from scratch 1`] = `
"{
  "version": 1,
  "projects": {
    "THEPROJECT": {
      "projectType": "application",
      "root": "PROJECTROOT",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/THEPROJECT"
          }
        },
        "deploy": {
          "builder": "angular-cli-ghpages:deploy"
        }
      }
    },
    "OTHERPROJECT": {
      "projectType": "application",
      "root": "PROJECTROOT",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/OTHERPROJECT"
          }
        }
      }
    }
  }
}"
`;

exports[`ng-add generating files overrides existing files 1`] = `
"{
  "version": 1,
  "projects": {
    "THEPROJECT": {
      "projectType": "application",
      "root": "PROJECTROOT",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/THEPROJECT"
          }
        },
        "deploy": {
          "builder": "angular-cli-ghpages:deploy"
        }
      }
    },
    "OTHERPROJECT": {
      "projectType": "application",
      "root": "PROJECTROOT",
      "architect": {
        "build": {
          "options": {
            "outputPath": "dist/OTHERPROJECT"
          }
        },
        "deploy": {
          "builder": "angular-cli-ghpages:deploy"
        }
      }
    }
  }
}"
`;
